sherpa/fit.py

```diff
@@ -1119,29 +1235,63 @@ def fit(self, outfile=None, clobber=False, numcores=1):
         #       investigated if it is possible to pass that check
         #       but fail the following.
         #
-        if not iterable(dep) or len(dep) == 0:
+        if not np.iterable(dep) or len(dep) == 0:
             raise FitErr('nobins')
 
-        if ((iterable(staterror) and 0.0 in staterror) and
+        if ((np.iterable(staterror) and 0.0 in staterror) and
                 isinstance(self.stat, Chi2) and
                 type(self.stat) != Chi2 and
                 type(self.stat) != Chi2ModVar):
             raise FitErr('binhas0')
 
         init_stat = self.calc_stat()
-        # output = self.method.fit ...
-        tmp = self._iterfit._get_callback(outfile, clobber)
-        output = self._iterfit.fit(tmp,
-                                   self.model.thawedpars,
-                                   self.model.thawedparmins,
-                                   self.model.thawedparmaxes)
-        # LevMar always calculate chisquare, so call calc_stat
-        # just in case statistics is something other then chisquare
-        self.model.thawedpars = output[1]
-        tmp = list(output)
-        tmp[2] = self.calc_stat()
-        output = tuple(tmp)
-        # end of the gymnastics 'cause one cannot write to a tuple
+
+        # Using close_on_exit is not ideal but the output is optional,
+        # and it depends on what the input argument was, so it's
+        # awkward to do this with a context manager.
+        #
+        close_on_exit = True
```

Comment:
One of the big advantages of a context manager is that it'll close the file is something goes wrong. In normal operations, `close_on_exit` is equivalent, but when something goes wrong, it's not. However, context managers a fancy way to call "close". We could just wrap it into an old fashioned "try" block:
```Python
try:
   ... do stuff with cb and hf, maybe from line 1279 on ...
finally:
    if close_on_exit:
        fh.close()
```

This is a suggestion to discuss how to handle this situation not a request to do exactly this.

Reply:
This should now be addressed, but I'll leave the issue open just to check.